lib/maths/CBoostedTreeFactory.cc

```diff
@@ -315,37 +293,228 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) con
+        for (std::size_t i = 0; i < INITIAL_REGULARIZER_SEARCH_ITERATIONS; ++i) {
+            residualMoments.add(testLosses[i] - leastSquaresQuadraticTestLoss.predict(
+                                                    static_cast<double>(i)));
+        }
+        double sigma{std::sqrt(CBasicStatistics::variance(residualMoments))};
+        double logScaleAtThreeSigma{std::sqrt(3.0 * sigma / curvature)};
+        if (logBestRegularizerScale - logScaleAtThreeSigma >= leftEndpoint) {
+            // These are scales > bestRegularizerScale hence 1 / multiplier.
+            interval(MAX_REGULARIZER_INDEX) = std::min(
+                std::pow(1.0 / multiplier, logScaleAtThreeSigma), MAX_REGULARIZER_SCALE);
+        }
+        if (logBestRegularizerScale + logScaleAtThreeSigma <= rightEndpoint) {
+            interval(MIN_REGULARIZER_INDEX) = std::max(
+                std::pow(multiplier, logScaleAtThreeSigma), MIN_REGULARIZER_SCALE);
+        }
+    }
+    interval *= bestRegularizerScale;
+
+    return successful ? TOptionalVector{interval} : TOptionalVector{};
+}
+
 CBoostedTreeFactory CBoostedTreeFactory::constructFromParameters(std::size_t numberThreads,
                                                                  TLossFunctionUPtr loss) {
-    return {numberThreads, std::move(loss)};
+    return {false, numberThreads, std::move(loss)};
 }
 
-CBoostedTreeFactory::TBoostedTreeUPtr
-CBoostedTreeFactory::constructFromString(std::istream& jsonStringStream,
-                                         core::CDataFrame& frame,
-                                         TProgressCallback recordProgress,
-                                         TMemoryUsageCallback recordMemoryUsage,
-                                         TTrainingStateCallback recordTrainingState) {
+CBoostedTreeFactory CBoostedTreeFactory::constructFromString(std::istream& jsonStringStream) {
+    CBoostedTreeFactory result{true, 1, nullptr};
     try {
-        TBoostedTreeUPtr treePtr{new CBoostedTree{
-            frame, std::move(recordProgress), std::move(recordMemoryUsage),
-            std::move(recordTrainingState), TBoostedTreeImplUPtr{new CBoostedTreeImpl{}}}};
         core::CJsonStateRestoreTraverser traverser(jsonStringStream);
-        if (treePtr->acceptRestoreTraverser(traverser) == false || traverser.haveBadState()) {
+        if (result.m_TreeImpl->acceptRestoreTraverser(traverser) == false ||
+            traverser.haveBadState()) {
             throw std::runtime_error{"failed to restore boosted tree"};
         }
-        frame.resizeColumns(treePtr->m_Impl->m_NumberThreads,
-                            frame.numberColumns() +
-                                treePtr->m_Impl->numberExtraColumnsForTrain());
-        return treePtr;
     } catch (const std::exception& e) {
-        HANDLE_FATAL(<< "Input error: '" << e.what() << "'. Check logs for more details.");
+        throw std::runtime_error{std::string{"Input error: '"} + e.what() + "'"};
```

Comment:
Note I changed this to throw. `HANDLE_FATAL` will abort the program, but this is recoverable since we can just start again. The exception is already handled in the runner classes in the api library.